core/src/main/scala/kafka/log/Log.scala

```diff
@@ -403,21 +456,52 @@ class Log(@volatile var dir: File,
+
+      // Recover this swap segment. If this swap segment contains offsets that overflow the index (KAFKA-6264),
+      // recoverSegment wil throw an IndexOffsetOverflowException. We currently do not have a good way of dealing with
+      // this situation so we let this exception propagate all the way up to KafkaServer#startup which will log
+      // this as a fatal exception and will cause the broker to shut down. This scenario is expected to be extremely rare
+      // in practice, and manual intervention might be required to get out of the situation.
       recoverSegment(swapSegment)
-      val oldSegments = logSegments(swapSegment.baseOffset, swapSegment.readNextOffset)
-      replaceSegments(swapSegment, oldSegments.toSeq, isRecoveredSwapFile = true)
+
+      var oldSegments = logSegments(swapSegment.baseOffset, swapSegment.readNextOffset)
+
+      // We create swap files for two cases: (1) Log cleaning where multiple segments are merged into one, and
+      // (2) Log splitting where one segment is split into multiple.
+      // Both of these mean that the resultant swap segments be composed of the original set, i.e. the swap segment
+      // must fall within the range of existing segment(s). If we cannot find such a segment, it means the deletion
+      // of that segment was successful. In such an event, we should simply rename the .swap to .log without having to
+      // do a replace with an existing segment.
+      if (oldSegments.nonEmpty) {
+        val start = oldSegments.head.baseOffset
+        val end = oldSegments.last.readNextOffset
+        if (!(swapSegment.baseOffset >= start && swapSegment.baseOffset <= end))
+          oldSegments = List()
+      }
+
+      replaceSegments(Seq(swapSegment), oldSegments.toSeq, isRecoveredSwapFile = true)
     }
   }
 
   // Load the log segments from the log files on disk and return the next offset
   // This method does not need to convert IOException to KafkaStorageException because it is only called before all logs are loaded
+  @throws(classOf[IndexOffsetOverflowException])
   private def loadSegments(): Long = {
     // first do a pass through the files in the log directory and remove any temporary files
     // and find any interrupted swap operations
     val swapFiles = removeTempFilesAndCollectSwapFiles()
 
-    // now do a second pass and load all the log and index files
-    loadSegmentFiles()
+    // Now do a second pass and load all the log and index files.
+    // We might encounter legacy log segments with offset overflow (KAFKA-6264). Such segments would be split by
+    // Log#recoverSegment operation. When this happens, restart loading segment files from scratch. Theoretically, it is
+    // possible that we see IndexOffsetOverflowException as many times as the number of log segments.
+    var maxTries = 1
+    for (file <- dir.listFiles)
+      if (isLogFile(file))
+        maxTries += 1
+    Log.maybeRetryOnOffsetOverflow(maxTries) {
```

Comment:
The function `loadSegmentFiles` mutates the internal state of the log. Some segments may get loaded on the first attempt before encountering an overflow problem. After the retry, they will get loaded again. What ensures that the segments loaded in the first pass are closed properly?

Reply:
I think the only thing we need to do is close the files that were opened. I added some exception handling code for this - let me know if it looks.